File: src/test/java/org/vanautrui/languages/dragoncompiler/parsing/astnodes/nonterminals/statements/controlflow/ReturnStatementNodeTest.java
package org.vanautrui.languages.dragoncompiler.parsing.astnodes.nonterminals.statements.controlflow;

import org.junit.Test;
import org.vanautrui.languages.dragoncompiler.lexing.utils.TokenList;
import org.vanautrui.languages.dragoncompiler.lexing.tokens.*;
import org.vanautrui.languages.dragoncompiler.parsing.astnodes.nonterminal.statements.controlflow.ReturnStatementNode;

public class ReturnStatementNodeTest {

    @Test
    public void test1() throws Exception {
        TokenList list = new TokenList();
        list.add(new KeywordToken("return"));

        list.add(new SymbolToken("("));

	list.add(new OperatorToken("-"));
        list.add(new IntegerNonNegativeConstantToken(5));

        list.add(new SymbolToken(")"));

        list.add(new SymbolToken(";"));

        ReturnStatementNode returnStatementNode = new ReturnStatementNode(list);
    }

    @Test
    public void test2() throws Exception {
        TokenList list = new TokenList();
        list.add(new KeywordToken("return"));

        list.add(new SymbolToken("("));

	list.add(new OperatorToken("-"));
        list.add(new IntegerNonNegativeConstantToken(5));

        list.add(new SymbolToken(")"));

        list.add(new OperatorToken("*"));

        list.add(new IdentifierToken("n"));

        list.add(new SymbolToken(";"));

        ReturnStatementNode returnStatementNode = new ReturnStatementNode(list);

        System.out.println(returnStatementNode.toSourceCode());
    }

    @Test
    public void test3() throws Exception {
        TokenList list = new TokenList();
        list.add(new KeywordToken("return"));

        list.add(new IdentifierToken("arr"));

        list.add(new SymbolToken("["));
        list.add(new IntegerNonNegativeConstantToken(0));

        list.add(new SymbolToken("]"));

        list.add(new SymbolToken(";"));

        ReturnStatementNode returnStatementNode = new ReturnStatementNode(list);
    }
}

File: dragon-parser/src/test/astnodes/nonterminals/statements/controlflow/ReturnStatementNodeTest.cpp
package org.vanautrui.languages.test.astnodes.nonterminals.statements.controlflow;

import org.vanautrui.languages.compiler.lexing.tokens.IdentifierToken;
import org.vanautrui.languages.compiler.lexing.tokens.IntegerNonNegativeConstantToken;
import org.vanautrui.languages.compiler.lexing.tokens.OperatorToken;
import org.vanautrui.languages.compiler.lexing.tokens.no_members.*;
import org.vanautrui.languages.compiler.lexing.tokens.no_members.keywords.ReturnToken;
import org.vanautrui.languages.compiler.lexing.utils.TokenList;
import org.vanautrui.languages.compiler.parsing.astnodes.nonterminal.statements.controlflow.ReturnStatementNode;

import java.nio.file.Paths;

public final class ReturnStatementNodeTest {

	public static int test1() {
		try {
			final TokenList list = new TokenList(Paths.get("/dev/null"));
			list.add(new ReturnToken());

			list.add(new LParensToken());

			list.add(new OperatorToken("-"));
			list.add(new IntegerNonNegativeConstantToken(5));

			list.add(new RParensToken());

			list.add(new SemicolonToken());

			final ReturnStatementNode returnStatementNode = new ReturnStatementNode(list);

			return 1;
		}catch (Exception e){
			return 0;
		}
	}

	public static int test2() throws Exception {
		try {
			final TokenList list = new TokenList(Paths.get("/dev/null"));
			list.add(new ReturnToken());

			list.add(new LParensToken());

			list.add(new OperatorToken("-"));

			list.add(new IntegerNonNegativeConstantToken(5));

			list.add(new RParensToken());

			list.add(new OperatorToken("*"));

			list.add(new IdentifierToken("n"));

			list.add(new SemicolonToken());

			final ReturnStatementNode returnStatementNode = new ReturnStatementNode(list);

			return 1;
		}catch (Exception e){
			return 0;
		}
	}

	public static int test3() throws Exception {

		try {
			final TokenList list = new TokenList(Paths.get("/dev/null"));

			list.add(new ReturnToken());

			list.add(new IdentifierToken("arr"));

			list.add(new LBracketToken());
			list.add(new IntegerNonNegativeConstantToken(0));

			list.add(new RBracketToken());

			list.add(new SemicolonToken());

			final ReturnStatementNode returnStatementNode = new ReturnStatementNode(list);
			return 1;
		}catch (Exception e){
			return 0;
		}
	}


}
